crates/sui-core/src/authority/authority_store_pruner.rs

```diff
@@ -144,6 +161,111 @@ impl AuthorityStorePruner {
+                    &next_digest,
+                )?;
+                perpetual_batch.delete_range(
+                    &perpetual_db.executed_effects,
+                    &transaction_digest,
+                    &next_digest,
+                )?;
+                perpetual_batch.delete_range(
+                    &perpetual_db.executed_transactions_to_checkpoint,
+                    &transaction_digest,
+                    &next_digest,
+                )?;
+            }
+        }
+
+        let mut events_to_prune = vec![];
+        for effects in effects_to_prune {
+            let effects_digest = effects.digest();
+            if let Some(next_digest) = effects.digest().next_lexicographical() {
+                perpetual_batch.delete_range(
+                    &perpetual_db.effects,
+                    &effects_digest,
+                    &next_digest,
+                )?;
+            }
+            events_to_prune.push(effects.events_digest().cloned());
+        }
+
+        for event_digest in events_to_prune.iter().flatten() {
+            if let Some(next_digest) = event_digest.next_lexicographical() {
+                perpetual_batch.delete_range(
+                    &perpetual_db.events,
+                    &(*event_digest, 0),
+                    &(next_digest, 0),
+                )?;
+            }
+        }
+
+        let mut checkpoints_batch = checkpoint_db.certified_checkpoints.batch();
+
+        for checkpoint_content in checkpoint_content_to_prune {
+            let content_digest = *checkpoint_content.digest();
+            if let Some(next_digest) = content_digest.next_lexicographical() {
+                checkpoints_batch.delete_range(
+                    &checkpoint_db.checkpoint_content,
+                    &content_digest,
+                    &next_digest,
+                )?;
+                checkpoints_batch.delete_range(
+                    &checkpoint_db.checkpoint_sequence_by_contents_digest,
+                    &content_digest,
+                    &next_digest,
+                )?;
+            }
+        }
+        for (seq_number, checkpoint_digest) in checkpoints_to_prune {
+            if let Some(next_digest) = checkpoint_digest.next_lexicographical() {
+                checkpoints_batch.delete_range(
+                    &checkpoint_db.checkpoint_by_digest,
+                    &checkpoint_digest,
+                    &next_digest,
+                )?;
+            }
+            checkpoints_batch.delete_range(
+                &checkpoint_db.certified_checkpoints,
+                &seq_number,
+                &(seq_number + 1),
+            )?;
+        }
+        checkpoints_batch.insert_batch(
+            &checkpoint_db.watermarks,
+            [(
+                &CheckpointWatermark::HighestPruned,
+                &(checkpoint_number, CheckpointDigest::random()),
+            )],
+        )?;
+
+        perpetual_batch.write()?;
+        checkpoints_batch.write()?;
```

Comment:
Just an observation here: Even though we are committing these write batches we have to be prepared for a world (and probably fix our code to account for it) where these deletes will *appear* to take affect out of order and in complete random order depending on when compaction gets triggered on these tables and databases. This is because of `delete_range` with ignore set to true